File: CodingTest/Python/solution15.py
# https://school.programmers.co.kr/learn/courses/30/lessons/42579
def solution(genres, plays):
    answer = []
    
    tables = {}
    
    for i, genre in enumerate(genres):
        if genre not in tables.keys():
            tables[genre] = [plays[i], []]
            tables[genre][1].append([i, plays[i]])
        else:
            tables[genre][0] += plays[i]
            tables[genre][1].append([i, plays[i]])

    lists = list(tables.values())
    lists.sort(key=lambda x: x[0], reverse= True)
    for i in lists:
        for k, j in enumerate(sorted(i[1], key = lambda x : x[1], reverse= True)):
            if k > 1:
                break
            answer.append(j[0])
    
    return answer

File: CodingTest/Java/Solution15.java
// https://school.programmers.co.kr/learn/courses/30/lessons/42885
package CodingTest.Java;

import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;

public class Solution15 {
    public int solution(int[] people, int limit) {
        int answer = 0;
        Arrays.sort(people);
        Deque<Integer> queue = new LinkedList<>();
        for(int i = 0; i < people.length; i++) {
            queue.addLast(people[i]);
        }

        while (!queue.isEmpty()) {
            if (queue.size() == 1) {
                queue.removeLast();
                answer += 1;
            } else {
                if (queue.getFirst() + queue.getLast() > limit) {
                    queue.removeLast();
                    answer += 1;
                } else {
                    queue.removeFirst();
                    queue.removeLast();
                    answer += 1;
                }
            }
        }
        
        return answer;
    }
}